legion: Add 23.06.0, variants for UCX, max nodes, update CUDA version.

elliottslaughter opened the pull request.

elliottslaughter committed:
legion: Add 23.06.0, variants for UCX, max nodes, update CUDA version.

alecbcs commented on var/spack/repos/builtin/packages/legion/package.py:
```diff
@@ -46,10 +47,11 @@ class Legion(CMakePackage, ROCmPackage):
     depends_on("mpi", when="conduit=mpi")
-    depends_on("cuda@10.0:11.9", when="+cuda_unsupported_compiler")
-    depends_on("cuda@10.0:11.9", when="+cuda")
+    depends_on("cuda@10.0:12.2", when="+cuda_unsupported_compiler")
+    depends_on("cuda@10.0:12.2", when="+cuda")
```
@elliottslaughter do all versions of legion support up to `cuda@12.2` or should we be setting these bounds on a per version basis? (Will `legion@21.03.0` compile with `cuda@12.2`?)

elliottslaughter replied:
My understanding is that the existing Legion releases should work with CUDA 12.2, but only Legion 23.06.0 will work with 12.1 (due to a CUDA bug in 12.1). I'm not sure it's worth the effort to track version bounds at that granularity? Aside from CUDA bugs like this Legion usually works without any additional effort.

elliottslaughter replied:
An alternative would be just to say that we'll only add new CUDA versions for new Legion versions. That will be much more conservative (because usually those releases would work fine), but I'm not really interested in debugging compatibility issues on older Legions anyway, so it seems like the safer (and possibly saner) approach.

elliottslaughter replied:
@alecbcs Ok, I attempted to resolve this in the most recent version (taking tge more conservative approach in my second comment). Could you check to see if this looks reasonable?

elliottslaughter committed:
legion: Make newer CUDA versions dependent on newer Legion.

elliottslaughter committed:
legion: Update CUDA arch list so that we can stop tracking manually.

elliottslaughter commented:
Per the comment in https://github.com/StanfordLegion/legion/issues/1465#issuecomment-1631611106, I updated the CUDA arch list to borrow from `CudaPackage`. I also rebased to fix conflicts.

elliottslaughter commented:
Rebased again to resolve conflicts.

alecbcs reviewed (approved):
Looks good to me! Thanks for all the work sliming this down @elliottslaughter!
